src/app/icd/client/ICDRefreshKeyInfo.cpp

```diff
@@ -0,0 +1,147 @@
+    mpICDRefreshKeyInfo->GetICDClientInfo().start_icd_counter = responseData.ICDCounter;
+    mpICDRefreshKeyInfo->GetICDClientInfo().offset            = 0;
+    mpICDRefreshKeyInfo->GetICDClientStorage()->RemoveKey(mpICDRefreshKeyInfo->GetICDClientInfo());
+    mError = mpICDRefreshKeyInfo->GetICDClientStorage()->SetKey(mpICDRefreshKeyInfo->GetICDClientInfo(),
+                                                                mpICDRefreshKeyInfo->mNewKey.Span());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to set the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+        mpICDRefreshKeyInfo->GetCheckInDelegate()->OnKeyRefreshDone(mpICDRefreshKeyInfo->GetICDClientInfo(), mError);
+        return;
+    }
+
+    mError = mpICDRefreshKeyInfo->GetICDClientStorage()->StoreEntry(mpICDRefreshKeyInfo->GetICDClientInfo());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to store the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+        mpICDRefreshKeyInfo->GetCheckInDelegate()->OnKeyRefreshDone(mpICDRefreshKeyInfo->GetICDClientInfo(), mError);
+        return;
+    }
+
+    mpICDRefreshKeyInfo->GetCheckInDelegate()->OnCheckInComplete(mpICDRefreshKeyInfo->GetICDClientInfo());
+    mpICDRefreshKeyInfo->GetCheckInDelegate()->OnKeyRefreshDone(mpICDRefreshKeyInfo->GetICDClientInfo(), CHIP_NO_ERROR);
+}
+void ICDRefreshKeyInfo::Callback::OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError)
+{
+    mError = aError;
+    mpICDRefreshKeyInfo->GetCheckInDelegate()->OnKeyRefreshDone(mpICDRefreshKeyInfo->GetICDClientInfo(), mError);
+}
+
+void ICDRefreshKeyInfo::Callback::OnDone(chip::app::CommandSender * apCommandSender) {}
+
+CHIP_ERROR ICDRefreshKeyInfo::RegisterClientWithNewKey(Messaging::ExchangeManager & exchangeMgr,
+                                                       const SessionHandle & sessionHandle)
+{
+    using namespace Clusters::IcdManagement;
+    EndpointId endpointId = 0;
+    GroupId groupId       = 0;
+
+    mRegisterCommandSender.Emplace(&mCommandSenderDelegate, &exchangeMgr);
+
+    mCommandSenderDelegate.SetICDRefreshKeyInfo(this);
+
+    Commands::RegisterClient::Type registerClientCommand;
+    registerClientCommand.checkInNodeID    = mICDClientInfo.peer_node.GetNodeId();
+    registerClientCommand.monitoredSubject = mICDClientInfo.monitored_subject;
+    registerClientCommand.key              = mNewKey.Span();
+    auto commandPathParams =
+        CommandPathParams(endpointId, groupId, Id, Commands::RegisterClient::Id, (CommandPathFlags::kEndpointIdValid));
+
+    ReturnErrorOnFailure(mRegisterCommandSender.Value().AddRequestData(commandPathParams, registerClientCommand));
+    ReturnErrorOnFailure(mRegisterCommandSender.Value().SendCommandRequest(sessionHandle));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ICDRefreshKeyInfo::EstablishSessionToPeer()
```

Comment:
You may mimic TestRead to create the e2e test for Re-registration test from this function.

Reply:
Sure. Will add test as a follow-up PR.